File: python/enrich.py
import json
import re
from concurrent.futures import ThreadPoolExecutor, as_completed

import requests
from bs4 import BeautifulSoup

INPUT_FILE = "episodes.json"

HEADERS = {
    "User-Agent": "Mozilla/5.0"
}


def extract_metadata(page_url):

    response = requests.get(
        page_url,
        headers=HEADERS,
        timeout=15
    )

    response.raise_for_status()

    soup = BeautifulSoup(
        response.text,
        "html.parser"
    )

    data = {}

    # -----------------------------------------
    # EPISODE HEADING
    # -----------------------------------------

    heading = soup.find("h1")

    if heading:

        title = heading.get_text(
            " ",
            strip=True
        )

        # AniZone currently gives:
        # "- Episode 1"
        data["title"] = title.lstrip("- ").strip()

    # -----------------------------------------
    # FIND MAIN EPISODE INFORMATION
    # -----------------------------------------

    # Find the h1 and use the following
    # content rather than random <p> tags.
    #
    # This prevents the site's ad message
    # from becoming the summary.

    if heading:

        parent = heading.parent

        if parent:

            text = parent.get_text(
                "\n",
                strip=True
            )

            lines = [
                line.strip()
                for line in text.splitlines()
                if line.strip()
            ]

            # ---------------------------------
            # TYPE
            # ---------------------------------

            for value in (
                "Regular",
                "Special"
            ):

                if value in lines:

                    data["type"] = value
                    break

            # ---------------------------------
            # DATE
            # ---------------------------------

            for line in lines:

                match = re.search(
                    r"\d{4}-\d{2}-\d{2}",
                    line
                )

                if match:

                    data["air_date"] = (
                        match.group(0)
                    )

                    break

    # -----------------------------------------
    # SUMMARY
    # -----------------------------------------

    # AniZone's actual episode summary is the
    # paragraph after the episode information.
    #
    # We specifically reject the site's
    # ad-blocker message.

    paragraphs = soup.find_all("p")

    for paragraph in paragraphs:

        summary = paragraph.get_text(
            " ",
            strip=True
        )

        if not summary:
            continue

        if (
            "intrusive ads" in
            summary.lower()
        ):
            continue

        if (
            "disable your adblock" in
            summary.lower()
        ):
            continue

        if len(summary) >= 40:

            data["summary"] = summary
            break

    # -----------------------------------------
    # PAGE TEXT
    # -----------------------------------------

    lines = [
        line.strip()
        for line in soup.get_text(
            "\n",
            strip=True
        ).splitlines()
        if line.strip()
    ]

    # -----------------------------------------
    # DURATION
    # -----------------------------------------

    for i, line in enumerate(lines):

        if line == "Duration:":

            if i + 1 < len(lines):

                value = lines[i + 1]

                if re.match(
                    r"^\d+:\d+$",
                    value
                ):

                    data["duration"] = value

    # -----------------------------------------
    # SOURCE
    # -----------------------------------------

    for i, line in enumerate(lines):

        if line == "Source:":

            if i + 1 < len(lines):

                data["source"] = (
                    lines[i + 1]
                )

    # -----------------------------------------
    # AUDIO
    # -----------------------------------------

    for i, line in enumerate(lines):

        if line == "Audio:":

            if i + 1 < len(lines):

                data["audio"] = (
                    lines[i + 1]
                )

    # -----------------------------------------
    # SOFTSUB
    # -----------------------------------------

    for i, line in enumerate(lines):

        if line == "Softsub:":

            if i + 1 < len(lines):

                data["softsub"] = (
                    lines[i + 1]
                )

    return data


def fetch_episode(item):

    episode, episode_data = item

    try:

        metadata = extract_metadata(
            episode_data["page"]
        )

        return episode, metadata, None

    except Exception as error:

        return episode, {}, str(error)


def main():

    with open(
        INPUT_FILE,
        "r",
        encoding="utf-8"
    ) as file:

        episodes = json.load(file)

    total = len(episodes)

    print(
        f"Fetching {total} episodes..."
    )

    # 20 simultaneous requests.
    # Much faster than sequential fetching.

    with ThreadPoolExecutor(
        max_workers=20
    ) as executor:

        futures = [
            executor.submit(
                fetch_episode,
                item
            )
            for item in episodes.items()
        ]

        completed = 0

        for future in as_completed(futures):

            episode, metadata, error = (
                future.result()
            )

            completed += 1

            if error:

                print(
                    f"[{completed}/{total}] "
                    f"{episode}: ERROR {error}"
                )

                continue

            episodes[episode].update(
                metadata
            )

            print(
                f"[{completed}/{total}] "
                f"{episode}: ✓"
            )

    with open(
        INPUT_FILE,
        "w",
        encoding="utf-8"
    ) as file:

        json.dump(
            episodes,
            file,
            indent=2,
            ensure_ascii=False
        )

    print("\nDone.")


if __name__ == "__main__":
    main()

File: rust/src/bin/enrich.rs
use std::collections::BTreeMap;
use std::sync::mpsc;
use std::thread;
use std::time::Duration;

use anyhow::{Context, Result};
use regex::Regex;
use scraper::{Html, Selector};

use dbz::{load_episodes, save_episodes};

fn extract_metadata(page_url: &str) -> Result<BTreeMap<String, String>> {
    let client = reqwest::blocking::Client::new();

    let response = client
        .get(page_url)
        .header("User-Agent", "Mozilla/5.0")
        .timeout(Duration::from_secs(15))
        .send()
        .context("request failed")?;

    if !response.status().is_success() {
        anyhow::bail!("non-success status: {}", response.status());
    }

    let html = response.text().context("failed to read body")?;

    let document = Html::parse_document(&html);

    let mut data: BTreeMap<String, String> = BTreeMap::new();

    // -----------------------------------------
    // EPISODE HEADING
    // -----------------------------------------

    let h1_selector = Selector::parse("h1").unwrap();

    if let Some(heading) = document.select(&h1_selector).next() {
        let title = heading.text().collect::<Vec<_>>().join(" ").trim().to_string();
        // AniZone currently gives: "- Episode 1"
        data.insert("title".to_string(), title.trim_start_matches("- ").trim().to_string());

        // -----------------------------------------
        // FIND MAIN EPISODE INFORMATION
        // -----------------------------------------

        if let Some(parent) = heading.parent() {
            let parent = scraper::ElementRef::wrap(parent);
            let text = parent
                .map(|p| p.text().collect::<Vec<_>>().join("\n"))
                .unwrap_or_default();
            let lines: Vec<String> = text
                .lines()
                .map(|l| l.trim().to_string())
                .filter(|l| !l.is_empty())
                .collect();

            // TYPE
            for value in ["Regular", "Special"] {
                if lines.iter().any(|l| l == value) {
                    data.insert("type".to_string(), value.to_string());
                    break;
                }
            }

            // DATE
            let date_re = Regex::new(r"\d{4}-\d{2}-\d{2}").unwrap();
            for line in &lines {
                if let Some(m) = date_re.find(line) {
                    data.insert("air_date".to_string(), m.as_str().to_string());
                    break;
                }
            }
        }
    }

    // -----------------------------------------
    // SUMMARY
    // -----------------------------------------

    let p_selector = Selector::parse("p").unwrap();

    for paragraph in document.select(&p_selector) {
        let summary = paragraph.text().collect::<Vec<_>>().join(" ").trim().to_string();

        if summary.is_empty() {
            continue;
        }

        let lower = summary.to_lowercase();
        if lower.contains("intrusive ads") || lower.contains("disable your adblock") {
            continue;
        }

        if summary.chars().count() >= 40 {
            data.insert("summary".to_string(), summary);
            break;
        }
    }

    // -----------------------------------------
    // PAGE TEXT (for duration/source/audio/softsub)
    // -----------------------------------------

    let lines: Vec<String> = document
        .root_element()
        .text()
        .collect::<Vec<_>>()
        .join("\n")
        .lines()
        .map(|l| l.trim().to_string())
        .filter(|l| !l.is_empty())
        .collect();

    let duration_re = Regex::new(r"^\d+:\d+$").unwrap();

    for (i, line) in lines.iter().enumerate() {
        if line == "Duration:" {
            if let Some(value) = lines.get(i + 1) {
                if duration_re.is_match(value) {
                    data.insert("duration".to_string(), value.clone());
                }
            }
        }
    }

    for (i, line) in lines.iter().enumerate() {
        if line == "Source:" {
            if let Some(value) = lines.get(i + 1) {
                data.insert("source".to_string(), value.clone());
            }
        }
    }

    for (i, line) in lines.iter().enumerate() {
        if line == "Audio:" {
            if let Some(value) = lines.get(i + 1) {
                data.insert("audio".to_string(), value.clone());
            }
        }
    }

    for (i, line) in lines.iter().enumerate() {
        if line == "Softsub:" {
            if let Some(value) = lines.get(i + 1) {
                data.insert("softsub".to_string(), value.clone());
            }
        }
    }

    Ok(data)
}

fn fetch_episode(
    episode: String,
    episode_data: &dbz::Episode,
) -> (String, BTreeMap<String, String>, Option<String>) {
    let page = match &episode_data.page {
        Some(p) => p.clone(),
        None => return (episode, BTreeMap::new(), Some("no page URL".to_string())),
    };

    match extract_metadata(&page) {
        Ok(metadata) => (episode, metadata, None),
        Err(e) => (episode, BTreeMap::new(), Some(e.to_string())),
    }
}

fn main() -> Result<()> {
    let mut episodes = load_episodes()?;

    let total = episodes.len();

    println!("Fetching {total} episodes...");

    // 20 simultaneous requests.
    let (tx, rx) = mpsc::channel();

    let mut handles = Vec::new();

    for (number, data) in episodes.iter() {
        let tx = tx.clone();
        let number = number.clone();
        let data = data.clone();
        handles.push(thread::spawn(move || {
            let result = fetch_episode(number.clone(), &data);
            let _ = tx.send((number, result));
        }));
    }

    drop(tx);

    let mut completed = 0;

    for received in rx {
        let (number, (_, metadata, error)) = received;

        completed += 1;

        if let Some(err) = error {
            println!("[{completed}/{total}] {number}: ERROR {err}");
            continue;
        }

        if let Some(episode) = episodes.get_mut(&number) {
            episode.title = metadata.get("title").cloned();
            episode.kind = metadata.get("type").cloned();
            episode.air_date = metadata.get("air_date").cloned();
            episode.duration = metadata.get("duration").cloned();
            episode.source = metadata.get("source").cloned();
            episode.audio = metadata.get("audio").cloned();
            episode.softsub = metadata.get("softsub").cloned();
            episode.summary = metadata.get("summary").cloned();
        }

        println!("[{completed}/{total}] {number}: ✓");
    }

    for handle in handles {
        let _ = handle.join();
    }

    save_episodes(&episodes)?;

    println!("\nDone.");

    Ok(())
}